Add extra and gifted execution minutes

tw4l opened the pull request:
Fixes #1358 

Could use a good bit of additional manual testing, edge cases/fuzz testing welcomed (e.g. this has not been tested across months yet)! And very open to any suggestions to simplify the implementation on backend and/or frontend.

## Overview

- Adds `extraExecMinutes` and `giftedExecMinutes` org quotas, which are not reset monthly but a one-time update-able bucket
- Adds `quotaUpdate` field to `Organization` to track when quotas were updated with timestamp
- Adds `extraExecMinutesAvailable` and `giftedExecMinutesAvailable` fields to `Organization` to help with tracking available time left (includes tested migration to initialize these to 0)
- Modifies org backend to track time across multiple categories, using monthly exec time first, then gifted, then extra
- Updates Dashboard crawling meter to include all types of execution time if `extraExecMinutes` and/or `giftedExecMinutes` are set above 0
- Updates Dashboard Usage History table to include all types of execution time
- Adds backend nightly test to check handling of quotas and execution time

## Screenshots

#### Only monthly quota set, no execution time used yet

<img width="376" alt="Screen Shot 2023-11-15 at 1 20 50 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/2a038d76-c8dc-403d-906a-7081af5db75b">

#### Only monthly quota set, with some execution time used

<img width="386" alt="Screen Shot 2023-11-15 at 1 24 15 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/8b14a623-78b8-49f1-88f8-d695691f2192">

#### All three quotas set, no execution time used yet

<img width="380" alt="Screen Shot 2023-11-15 at 2 05 32 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/84b0b0ae-a520-413b-889d-1e8472b7ebaf">

#### All three quotas set, some execution time used

<img width="1189" alt="Screen Shot 2023-11-15 at 2 27 24 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/f2575602-8bc6-4633-8d61-3190dbccf887">

<img width="380" alt="Screen Shot 2023-11-15 at 2 43 30 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/4104a20a-7322-405a-87da-72418a2e9845">

<img width="383" alt="Screen Shot 2023-11-15 at 2 47 22 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/198278db-07cc-4aa7-87ec-ae62e3a5a5e4">

<img width="1172" alt="Screen Shot 2023-11-15 at 2 59 35 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/5b7d7aae-2cb3-4b87-846f-7dc919983676">

#### Tooltips

<img width="388" alt="Screen Shot 2023-11-15 at 2 27 36 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/f7750d02-4232-420e-a837-acb13296d249">

<img width="387" alt="Screen Shot 2023-11-15 at 2 27 39 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/43e4c6fb-1c6a-4d7a-81e3-106ccb525600">

#### Quota reached

<img width="1229" alt="Screen Shot 2023-11-15 at 3 04 20 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/51e79594-17e6-42b6-83b8-db17bc3f78b8">

#### With storage quota also enabled

<img width="1236" alt="Screen Shot 2023-11-15 at 3 05 36 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/c112917f-26bf-43f8-aef4-8f3db935f2f7">

tw4l committed:
Implement extra and gifted execution minutes

- Add new quotas
- Add usage tracking and quota update history to Organization
- Add tracking of available seconds for extra and gifted  minutes
to Organization
- Modify routine to check if quota is reached to return False if
extra or gifted time remains (with flag to disable checking these
for internal purposes)
- Update inc_org_time_stats to update used and remaining extra and
gifted time on crawl completion
- Update frontend execution time meter and usage history

tw4l committed:
Add migration to initialize extra/gifted exec secs available

tw4l committed:
Typing fixes

tw4l committed:
Remove right border on MeterBar

tw4l committed:
Round execution minutes down for meter available value

tw4l committed:
Further meter data improvements - cap usage at quotas

tw4l committed:
Fix calculation of crawl seconds in inc_org_time_stats

tw4l committed:
Add passing backend nightly test

tw4l commented on backend/btrixcloud/orgs.py:
```diff
@@ -377,21 +415,30 @@ async def get_this_month_crawl_exec_seconds(self, oid: UUID) -> int:
         quota_reached = False
 
+        if include_extra:
+            gifted_seconds = await self.get_gifted_exec_secs_available(oid)
+            if gifted_seconds:
+                return False
+
+            extra_seconds = await self.get_extra_exec_secs_available(oid)
+            if extra_seconds:
+                return False
```
```suggestion
            if await self.get_gifted_exec_secs_available(oid):
                return False

            if await self.get_extra_exec_secs_available(oid):
                return False
```

tw4l commented on backend/btrixcloud/orgs.py:
```diff
@@ -434,6 +495,82 @@ async def inc_org_time_stats(self, oid, duration, is_exec_time=False):
+
+        # If we're still within our monthly base quota, nothing to do
+        if not monthly_base_quota_passed:
+            return
```
```suggestion

        if not await self.exec_mins_quota_reached(oid, include_extra=False):
            return
```

tw4l commented on backend/btrixcloud/orgs.py:
```diff
@@ -434,6 +495,78 @@ async def inc_org_time_stats(self, oid, duration, is_exec_time=False):
+        if not await self.exec_mins_quota_reached(oid, include_extra=False):
+            return
+
+        # If there's no extra or gifted time available, nothing to do
```
```suggestion
```

tw4l commented on backend/test_nightly/test_execution_minutes_quota.py:
```diff
@@ -72,6 +76,92 @@ def test_crawl_stopped_when_quota_reached(org_with_quotas, admin_auth_headers):
+        },
+    )
+    data = r.json()
+    assert data.get("updated") == True
```
```suggestion
    assert r.json()["updated"]
````

Shrinks99 commented:
Getting a runtime error on this branch, ran `yarn` to make sure I was up to date but still no success: 

```txt
ERROR
Expected a finite number
module.exports@webpack-internal:///./node_modules/pretty-ms/index.js:10:9
Dashboard/this.formatExecutionSeconds/<@webpack-internal:///./src/pages/org/dashboard.ts:58:196
when@webpack-internal:///./node_modules/lit-html/development/directives/when.js:11:24
Dashboard/this.formatExecutionSeconds@webpack-internal:///./src/pages/org/dashboard.ts:58:69
renderCrawlingMeter/<@webpack-internal:///./src/pages/org/dashboard.ts:452:28
when@webpack-internal:///./node_modules/lit-html/development/directives/when.js:11:91
renderCrawlingMeter@webpack-internal:///./src/pages/org/dashboard.ts:441:71
render/<@webpack-internal:///./src/pages/org/dashboard.ts:235:22
renderCard/<@webpack-internal:///./src/pages/org/dashboard.ts:503:107
when@webpack-internal:///./node_modules/lit-html/development/directives/when.js:11:24
renderCard@webpack-internal:///./src/pages/org/dashboard.ts:503:73
render@webpack-internal:///./src/pages/org/dashboard.ts:234:18
update@webpack-internal:///./node_modules/lit-element/development/lit-element.js:138:28
performUpdate@webpack-internal:///./node_modules/@lit/reactive-element/development/reactive-element.js:831:22
scheduleUpdate@webpack-internal:///./node_modules/@lit/reactive-element/development/reactive-element.js:766:21
__enqueueUpdate@webpack-internal:///./node_modules/@lit/reactive-element/development/reactive-element.js:739:29
```

---
In the screenshots the inner shadow which helps the user discern the difference between the filled and unfilled part of the bar graph appears to be missing from the coloured backgrounds.  Out of context, a completely unfilled graph might look more filled than we want it to without this!

<img width="1064" alt="Screenshot 2023-11-16 at 1 20 12 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/5672810/259cabb4-9f06-4424-a4d8-3149ea3b25b5">

---

For the tooltips, it looks like you've made discrete tooltips for each section (filled and unfilled) whereas in the mockups I had them as one so users could easily compare how many minutes in each section they had used?  Open to other opinions if you have a reason this method is better.

I think I should have made a comment about this in the Figma, sorry :(

<img width="500" alt="Screenshot 2023-11-16 at 1 40 09 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/5672810/7ec08a1d-c8a6-4f34-a396-ff8aa0d17c12">

tw4l commented:
> Getting a runtime error on this branch, ran `yarn` to make sure I was up to date but still no success: ...

Looking into it!


> In the screenshots the inner shadow which helps the user discern the difference between the filled and unfilled part of the bar graph appears to be missing from the coloured backgrounds. Out of context, a completely unfilled graph might look more filled than we want it to without this!

Ah, I missed that and you're entirely right! Will add in a follow-up commit.

> For the tooltips, it looks like you've made discrete tooltips for each section (filled and unfilled) whereas in the mockups I had them as one so users could easily compare how many minutes in each section they had used? Open to other opinions if you have a reason this method is better.
> 
> I think I should have made a comment about this in the Figma, sorry :(

No reason to be sorry! This was just a holdover from a previous iteration and I hadn't noticed the difference in the mockups! Happy to change

tw4l commented:
~~Blocked on https://github.com/webrecorder/browsertrix-cloud/pull/1386, which refactors a helper used extensively in this PR. Will wait for that to be merged first.~~

Update: merged and rebased

emma-sg commented on frontend/src/pages/org/dashboard.ts:
```diff
@@ -78,9 +87,12 @@ export class Dashboard extends LiteElement {
+      ${when(
+        humanize,
+        () => html` (${humanizeMilliseconds(seconds * 1000)})`
+      )} `;
```
This all got refactored in https://github.com/webrecorder/browsertrix-cloud/pull/1386 — would be worth refactoring this to use & improve those utils. Specifically, `humanizeExecutionSeconds` with `style` `"short"` would do what you're looking for here, I think! Happy to help there, if you like.

Would be happy to do the rename of 'humanize' -> 'format' though, if you prefer!

tw4l replied:
Now rebased! Thanks for the tip on `style` `"short"` :) Just working through some of the other issues now and then will mark for re-review!

tw4l committed:
Add missing closing paren

tw4l committed:
Add left margin to Available

tw4l committed:
Available -> Remaining

tw4l committed:
Add box shadow to remaining bars in meter

tw4l committed:
Fix test

tw4l committed:
Show same tooltip on used and remaining bars

tw4l committed:
Refactor meter with new btrix-divider-meter-bar component

Combines used and remaining bars for each type into one
component with a shared tooltip.

Uses traditional meter-bar instead if no extra or gifted time has
been set.

tw4l commented:
@Shrinks99 @emma-sg Branch has been updated to resolve review comments, including:

- Rebase on latest humanizeExecutionSeconds work
- Add box shadow for remaining time bars
- Fix yarn runtime error
- Display single tooltip for used/remaining sections of each type of execution time when either extra or gifted execution minute quotas are enabled

### Updated screenshots

#### No extra/gifted execution minute quotas (similar behavior as before)

<img width="384" alt="Screen Shot 2023-11-22 at 1 19 01 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/e0547efb-5bfb-4819-8501-e82f71dcdb77">

<img width="442" alt="Screen Shot 2023-11-22 at 1 19 06 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/5af348af-2a38-4a8d-89a6-048dac1863a1">

<img width="384" alt="Screen Shot 2023-11-22 at 1 19 11 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/55a30d22-9e14-435a-92f6-3d4fb860d49e">

#### Extra/gifted execution minute quotas added

<img width="382" alt="Screen Shot 2023-11-22 at 1 19 26 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/ff7dea7c-4939-48f3-80cb-1f094cc9e9ab">

(Single tooltip per type of execution minutes, displays if cursor is over used or remaining, centering on used if present)

<img width="469" alt="Screen Shot 2023-11-22 at 1 19 30 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/425216cb-048b-40f6-8eca-f16833e4b843">

<img width="385" alt="Screen Shot 2023-11-22 at 1 19 35 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/71d2a1c0-31a7-40a2-899d-f3feb17c6fad">

<img width="391" alt="Screen Shot 2023-11-22 at 1 19 41 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/c17104b3-d22c-40f9-91db-2547a1110d5c">

<img width="382" alt="Screen Shot 2023-11-22 at 1 34 29 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/d5d338f2-a7f3-4c13-abb1-9b03399dfa51">

<img width="384" alt="Screen Shot 2023-11-22 at 1 44 15 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/36766459-cb8a-4d32-85e1-138401609736">

tw4l committed:
Format code and add right border-radius if used < quota

tw4l commented:
And one additional tweak to add right border-radius for execution time used bars if there is still remaining time in that category:

<img width="378" alt="Screen Shot 2023-11-22 at 2 08 39 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/07d8cdc8-bbf8-4761-8bda-284b69d60a91">

<img width="375" alt="Screen Shot 2023-11-22 at 1 53 08 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/f0bd600e-6a35-49df-b9f0-13174bd5dfd2">

tw4l committed:
Use remainder for style width

tw4l committed:
Fill background in under right border-radius

tw4l commented:
Right border-radius now fixed so that the quota bar shows underneath without any white pixels:

<img width="381" alt="Screen Shot 2023-11-22 at 3 46 26 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/3152ef7c-c45f-4a9e-83cd-30a2dbbed613">

<img width="377" alt="Screen Shot 2023-11-22 at 3 51 26 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/ba0fa003-fb85-4ae9-aaad-a91c6e8e1e2a">

Close-up:

<img width="503" alt="Screen Shot 2023-11-22 at 3 54 06 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/9a1aefff-fe79-450b-8302-748487d5eefa">

This also made it easier to always center the tooltip on the execution type, rather than just on the used portion, e.g.:

<img width="235" alt="Screen Shot 2023-11-22 at 3 55 31 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/1e6d53ab-36fa-4a52-a5cc-0236e9fcdf9a">

emma-sg committed:
refactor humanizeExecutionSeconds options
and show short units when style is short

tw4l committed:
Fix regression that made all meter quota bars the same size

tw4l commented on frontend/src/utils/executionTimeFormatter.ts:
```diff
@@ -67,23 +67,36 @@ export function humanizeSeconds(
+  options?: {
+    /**
+     * When this is "long", the time in hours is also displayed
+     * @default "full"
```
```suggestion
     * @default "long"
```

tw4l committed:
Update docstring default

tw4l committed:
Remove errant prop

emma-sg reviewed (approved):
Sweet! Nicely done, all works well as far as I can tell with my poking around

ikreymer commented on backend/btrixcloud/orgs.py:
```diff
@@ -434,6 +497,82 @@ async def inc_org_time_stats(self, oid, duration, is_exec_time=False):
+                {
+                    "$inc": {
+                        f"giftedExecSeconds.{yymm}": gifted_secs_available,
+                        "giftedExecSecondsAvailable": -gifted_secs_available,
```
This should always be 0, right?

tw4l replied:
Yup, just seemed easier to do the updates in one inc rather than splitting them into an inc and a set, but maybe a set to 0 would be clearer here

ikreymer replied:
yeah, would be clearer, though could also leave a comment perhaps?

tw4l replied:
Changed to a `$set`

ikreymer commented on backend/btrixcloud/orgs.py:
```diff
@@ -434,6 +497,82 @@ async def inc_org_time_stats(self, oid, duration, is_exec_time=False):
+            monthly_quota_secs + gifted_secs_available
+        )
+        extra_secs_available = org.extraExecSecondsAvailable
+        if extra_secs_available:
```
hm, should this still be `secs_to_use = min(secs_still_over_quota, extra_secs_available)`
and then everywhere else should do `if secs_to_use` and all incremented values should also use the min?

tw4l replied:
Modified!

ikreymer commented:
Nice! Looks good overall, but noticed one issue:
When changing the quota for the current month, I noticed that the bar resets to the entire monthly quota being used.
See attachment:
<img width="1189" alt="Screenshot 2023-11-27 at 10 30 53 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/1015759/ef6b2f17-628a-4571-969f-065f7bd95079">

The actual usage is 3m 2s but the meter shows that all 20 mins have been used

tw4l committed:
Apply code review suggestions

- Explicitly set giftedExecSecondsAvailable to 0 if appropriate
- Use min of seconds still over quota or available extra seconds
for calculating extraExecSeconds usage

tw4l committed:
Format

tw4l commented:
> Nice! Looks good overall, but noticed one issue: When changing the quota for the current month, I noticed that the bar resets to the entire monthly quota being used. See attachment: <img alt="Screenshot 2023-11-27 at 10 30 53 PM" width="1189" src="https://private-user-images.githubusercontent.com/1015759/286154380-ef6b2f17-628a-4571-969f-065f7bd95079.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDExOTEyMjcsIm5iZiI6MTcwMTE5MDkyNywicGF0aCI6Ii8xMDE1NzU5LzI4NjE1NDM4MC1lZjZiMmYxNy02MjhhLTQ1NzEtOTY5Zi0wNjVmN2JkOTUwNzkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTEyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzExMjhUMTcwMjA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmVjMmI4YWMzYzgyMDJmOWRmZGNlYmQ1YTFjNWM2ODkxYTkzZmMwMGM2OThhZmJlMmQwZjBiMGE3MzdmMDliYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.PZkjrLB7fhN7iwOPGp0sQ5jfbAIFOvbzHOL-H1k9eEM">
> 
> The actual usage is 3m 2s but the meter shows that all 20 mins have been used

I haven't been able to reproduce this! Tried a few times with a few different orders of operation. Can you retrace your steps and see if the order of operations matters?

What does `crawlExecTime` in the org fetch in the network tab show? It could be that the crawl did go over what's shown in the chart below, e.g. if it was slow to stop after hitting a limit before you increased the amount of time? Would be good to know if it's the meter or chart or both that are inaccurate.

tw4l committed:
Clear available secs if gifted or extra time set to 0 on later update

tw4l committed:
Lowercase 'remaining' in dashboard

tw4l committed:
Code review revisions

- Make crawlExecSeconds total
- Add monthlyExecSeconds
- Add migration to copy previous crawlExecSeconds to monthly, since
other categories didn't exist before now
- Don't let quota updates result in negative time available, set to
0 instead when applicable

tw4l committed:
Linting fixes

tw4l committed:
Only show usage columns with data

tw4l committed:
Fix migration by using aggregation pipeline

ikreymer commented on backend/btrixcloud/orgs.py:
```diff
@@ -303,32 +303,31 @@ async def update_quotas(self, org: Organization, quotas: OrgQuotas):
 
         # Inc org available fields for extra/gifted execution time as needed
-        if quotas.extraExecMinutes:
+        if quotas.extraExecMinutes or quotas.extraExecMinutes == 0:
```
would it be clearer to make this `if quotas.extraExecMinutes is not None`?

ikreymer commented:
Let's update the logic so that monthlyExecSeconds <= monthly quota.
That means if there's no monthly quota, it is not incremented, and otherwise only incremented up to the quota, eg.
if there's a quota of 6000 seconds but actual shutdown takes 6030 seconds, crawlExecSeconds can go up to 6030 while monthlyExecSeconds will top at 6000.

tw4l committed:
Cap monthly, gifted, extra exec time at quotas

tw4l committed:
Fix crawl details execution seconds

tw4l commented:
> Let's update the logic so that monthlyExecSeconds <= monthly quota. That means if there's no monthly quota, it is not incremented, and otherwise only incremented up to the quota, eg. if there's a quota of 6000 seconds but actual shutdown takes 6030 seconds, crawlExecSeconds can go up to 6030 while monthlyExecSeconds will top at 6000.

Done!

tw4l committed:
Simplify is not None check

tw4l committed:
Fix typo

ikreymer reviewed (approved):
I think we can finally merge this! (@tw4l just want to make sure the PR message is still accurate, if any tweaks should be made before merging)
